qiskit_experiments/curve_analysis/curve_analysis.py

```diff
@@ -194,23 +194,12 @@ def _run_data_processing(
                     f"X value key {opt.x_key} is not defined in the circuit metadata."
                 ) from ex
             # Classify fit model
-            if len(self._models) == 1:
+            for m_id, (m_name, model_spec) in enumerate(self.options.data_subfit_map.items()):
+                if model_spec.items() <= metadata.items():
+                    break
+            else:
```

Comment:
> one model that gets reused for multiple series that are distinguished by their metadata and the filtering of the data_subfit_map dictionary. 

This is not correct because new fit model only works with the phase data, while what the `_run_data_processing` method generates is P1 data of (X, Y) quadrature. When I assign model_id (0, 1) to them, the fitter or initial guess subroutine may receive unexpected numbers because they filter the data with the index. We cannot discard X Y data because they are used to create figure at later time.

Reply:
This [1038bd8](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1243/commits/1038bd8b834fecb6ef98a56513347ae701a02b1e) is the alternative implementation, in which we let the curve analysis baseclass `_run_data_processing` classify the data based on the fit models, which yields false classification in this analysis. Then ramsey fitter subclass does re-classification in its format method (or it could overwrite `_run_data_processing` as well).

Reply:
I am having some trouble matching up your comments to mine and to the code changes, but I think the way you changed things is okay.

> one model that gets reused for multiple series that are distinguished by their metadata and the filtering of the data_subfit_map dictionary.

Here I was commenting on what could be possible and not specifically on what you were doing, so that was confusing. I meant that by matching the `data_subfit_map` the code could all using one model for multiple series. Like `RamseyXY` could use a single sinusoidal model and map the data to two series to fit X and Y. I didn't mean that here the X and Y data were getting mapped to a single model that fit their phase.

> This [1038bd8](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1243/commits/1038bd8b834fecb6ef98a56513347ae701a02b1e) is the alternative implementation, in which we let the curve analysis baseclass _run_data_processing classify the data based on the fit models, which yields false classification in this analysis. Then ramsey fitter subclass does re-classification in its format method

I thought the original PR also misclassified the XY series and then fixed it in `_format_data`?

Reply:
Regarding overriding `_run_data_processing`, let's review the current processing chain for the data:

1. Call `_run_analysis()` on `ExperimentData`
2. Call `CurveAnalysis._run_data_processing()` on `ExperimentData`
  i. Optional filtering of the data (I think only used by CrossResonanceHamiltonianAnalysis)
  ii. Map each `ExperimentData.data()` entry to a model ID.
  iii. Run the data processor on all `Experiment.data()` (was trained earlier).
3. Call `CurveAnalysis._format_data` on the `ScatterTable` returned by `_run_data_processing()`.
  + In the base class, this mostly just averages the data and labels it `formatted`.
4. Call `CurveAnalysis._run_curve_fit()` on the `ScatterTable` returned by `_format_data()`
  + This passes the full `ScatterTable` to `CurveData._generate_fit_guesses()`
  + This groups the data by model ID and passes the group yval, yerr, and xval to the model.

So one of the first steps is to map each `ExperimentData.data()` to a model ID. For the simple case where each measurement maps one to one to a model, this is fine, but here we see that it might not be general enough since we want to combine two measurements into one formatted data point for fitting. We could just override `_format_data` to produce the combined points as the formatted points but we want to keep the XY formatted  points as well for the plotting code. We solve this by cheating a little bit and assigning "n/a" model ID to the XY points because the fitting can skip that but the plotting goes by the model name instead of the model ID.

Some things to think about:

1. Assignment of model ID could be delayed until the formatted data rather than applied to the processed data. This is inconvenient to do because the mapping is based on the result metadata and it is better for the data frame to hold an integer column than to keep looking at dictionaries of metadata. As long as the mapping is like here where non-overlapping sets get mapped to one point it is clean. A case where overlapping sets of points got mapped to new fit curves would be more problematic.
2. Classification of curves to be plotted could be done separately from data to be fit. Then the XY data could be captured as formatted data into the scatter table without mapping it to a model ID.

Reply:
I did cleanup of the scatter table in #1319 . The point of refactoring is `model_name` -> `name` and `model_id` -> `class_id`. In the raw category data, the name and id doesn't need to be consistent with the fit model. The PR is rebased and the analysis class is reimplemented in [bdd83d1](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1243/commits/bdd83d1c65731f4ef3ebe485ba4282e71f10f136)